iceoryx2/src/port/details/outgoing_connections.rs

```diff
@@ -0,0 +1,396 @@
+            &self.message_type_details,
+            shm_pointer,
+            sample_size,
+        ))
+    }
+
+    pub(crate) fn borrow_sample(&self, offset: PointerOffset) -> (u64, usize) {
+        let segment_id = offset.segment_id();
+        let segment_state = &self.segment_states[segment_id.value() as usize];
+        let mut payload_size = segment_state.payload_size();
+        if segment_state.payload_size() == 0 {
+            payload_size = self.data_segment.bucket_size(segment_id);
+            segment_state.set_payload_size(payload_size);
+        }
+        (segment_state.borrow_sample(offset.offset()), payload_size)
+    }
+
+    pub(crate) fn retrieve_returned_samples(&self) {
+        for i in 0..self.len() {
+            if let Some(ref connection) = self.get(i) {
+                loop {
+                    match connection.sender.reclaim() {
+                        Ok(Some(ptr_dist)) => {
+                            self.release_sample(ptr_dist);
+                        }
+                        Ok(None) => break,
+                        Err(e) => {
+                            warn!(from self, "Unable to reclaim samples from connection {:?} due to {:?}. This may lead to a situation where no more samples will be delivered to this connection.", connection, e)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) fn release_sample(&self, offset: PointerOffset) {
+        if self.segment_states[offset.segment_id().value() as usize].release_sample(offset.offset())
+            == 1
+        {
+            unsafe {
+                self.data_segment.deallocate_bucket(offset);
+            }
+        }
+    }
+
+    fn remove_connection(&self, i: usize) {
+        if let Some(connection) = self.get(i) {
+            // # SAFETY: the receiver no longer exist, therefore we can
+            //           reacquire all delivered samples
+            unsafe {
+                connection
+                    .sender
+                    .acquire_used_offsets(|offset| self.release_sample(offset))
+            };
+
+            *self.get_mut(i) = None;
+        }
+    }
+
+    pub(crate) fn start_update_connection_cycle(&self) {
+        self.visitor.next_cycle();
+    }
+
+    pub(crate) fn update_connection<E: Fn(&Connection<Service>)>(
+        &self,
+        index: usize,
+        receiver_details: ReceiverDetails,
+        establish_new_connection_call: E,
+    ) -> Result<(), ZeroCopyCreationError> {
+        let create_connection = match self.get(index) {
+            None => true,
+            Some(connection) => {
+                let is_connected = connection.receiver_port_id == receiver_details.port_id;
+                if is_connected {
+                    self.visitor.visit(connection);
+                }
+                !is_connected
+            }
+        };
+
+        if create_connection {
+            match self.create(index, receiver_details) {
```

Comment:
The `incoming_connection.rs` file had a `self.prepare_connection_removal(index);` in between these two lines. Is it intentional not to have it here?

Reply:
Yes. The subscriber shall keep the connection open to publisher even when it is already removed - to receive remaining samples of that publisher.

On the publisher side you do not have such a problem for subscribers.

In the next PR those classes are renamed to understand the purpose better. Then they will be called sender/receiver.